monix-catnap-micrometer/src/main/scala/com/avast/sst/monix/catnap/micrometer/MicrometerCircuitBreakerMetrics.scala

```diff
@@ -0,0 +1,41 @@
+package com.avast.sst.monix.catnap.micrometer
+
+import java.util.concurrent.atomic.AtomicInteger
+
+import cats.effect.Sync
+import com.avast.sst.monix.catnap.CircuitBreakerMetrics
+import com.avast.sst.monix.catnap.CircuitBreakerMetrics.State
+import com.avast.sst.monix.catnap.CircuitBreakerMetrics.State.{Closed, HalfOpen, Open}
+import io.micrometer.core.instrument.MeterRegistry
+
+trait MicrometerCircuitBreakerMetrics[F[_]] extends CircuitBreakerMetrics[F]
+
+object MicrometerCircuitBreakerMetrics {
+
+  private[micrometer] class Impl[F[_]: Sync](name: String, meterRegistry: MeterRegistry) extends MicrometerCircuitBreakerMetrics[F] {
+
+    private val F = Sync[F]
+
+    private[this] val CircuitOpened = -1
+    private[this] val CircuitHalfOpened = 0
+    private[this] val CircuitClosed = 1
+
+    private val accepted = meterRegistry.counter(s"circuit-breaker.$name.accepted")
+    private val rejected = meterRegistry.counter(s"circuit-breaker.$name.rejected")
+    private val circuitState = meterRegistry.gauge[AtomicInteger](s"circuit-breaker.$name.state", new AtomicInteger(CircuitClosed))
+
+    override def increaseAccepted: F[Unit] = F.delay(accepted.increment())
+
+    override def increaseRejected: F[Unit] = F.delay(rejected.increment())
+
+    override def setState(state: State): F[Unit] = {
+      state match {
+        case Closed   => F.delay(circuitState.set(CircuitClosed))
+        case Open     => F.delay(circuitState.set(CircuitOpened))
+        case HalfOpen => F.delay(circuitState.set(CircuitHalfOpened))
+      }
+    }
+
+  }
+
+}
```

Comment:
I think
```scala
object MicrometerCircuitBreakerMetrics {
  private class MicrometerCircuitBreakerMetricsImpl[F[_]: Sync](name: String, meterRegistry: MeterRegistry) extends CircuitBreakerMetrics[F] {
    ...
  }
  def make[F[_]: Sync](name: String, meterRegistry: MeterRegistry): CircuitBreakerMetrics[F] = new MicrometerCircuitBreakerMetricsImpl(name, meterRegistry)
}
```
is just enough, there's no need to add more concepts ;) `MicrometerCircuitBreakerMetrics` as defined brings 0 value

Reply:
I still think modules should return "specific" thing instead of the most abstract thing. Question: if I remove the now unneeded trait will I be able to introduce it later in backwards-compatible way to allow some specialization of the returned instance? I am not sure... (e.g. let's say I want the returned object to provide the underlying `MeterRegistry`)

Reply:
> if I remove the now unneeded trait will I be able to introduce it later in backwards-compatible way to allow some specialization of the returned instance

Yep. Change `CircuitBreakerMetrics` in that case.
Or, in a case of a more radical change when changing it would be undesirable, introduce a new interface which inherits from `CircuitBreakerMetrics`, but only if you're in such situation. For example sake, we could name it `CircuitBreakerMetricsWithRegistry`. And then `object MicrometerCircuitBreakerMetrics` would then return an instance of `CircuitBreakerMetricsWithRegistry`.

In general, by returning an instance of an interface, you're also communicating an intent to the users which methods the are supposed to use/rely on, and which not. IMHO these design principles help users to see the intents of the designers without reading long documentation and also potential new contributors to this project.

Reply:
I still think it's worthwhile to have a `make` method to create an instance of `CircuitBreakerMetrics[F]` :wink:

Reply:
I don't understand. I think it's there, see `MicrometerCircuitBreakerMetricsModule`.

Reply:
currently we have:
```scala
// MicrometerCircuitBreakerMetrics.scala
object MicrometerCircuitBreakerMetrics {
  private[micrometer] class Impl[F[_]: Sync](name: String, meterRegistry: MeterRegistry) extends CircuitBreakerMetrics[F] {
    ...
  }
}

// MicrometerCircuitBreakerMetricsModule.scala
object MicrometerCircuitBreakerMetricsModule {
  def make[F[_]: Sync](name: String, meterRegistry: MeterRegistry): F[CircuitBreakerMetrics[F]] = {
    ...
  }
}
```

I propose something like this
```scala
// MicrometerCircuitBreakerMetrics.scala
object MicrometerCircuitBreakerMetrics {
  private class MicrometerCircuitBreakerMetricsImpl[F[_]: Sync](name: String, meterRegistry: MeterRegistry, circuitState: IORef[CircuitState]) extends CircuitBreakerMetrics[F] {
    ...
  }
  def make[F[_]: Sync](name: String, meterRegistry: MeterRegistry): F[CircuitBreakerMetrics[F]] = {
    for {
      circuitState <- IORef.make(CircuitState.CircuitClosed) // or however you create a new one
      cb = new MicrometerCircuitBreakerMetricsImpl(name, meterRegistry)
    } yield cb
  }
}
```

Reply:
Oh, I see now. The current structure is leftover after some refactorings. Needs to be changed.